### PR TITLE
add interest button and list favourited conferences on profile page

### PR DESCRIPTION
**Pull request by sophiambolah**

### CODE REVIEW

File: src/Conference/ConferenceDetails.jsx [hunk 1]
```diff
@@ -44,6 +44,14 @@ const ConferenceDetails = ({ conference, id, isLoading, error }) => {
     deleteConference(id);
   };
 
+  const expressInterest = async () => {
+    try {
+      await favouriteConference(conference, token);
```
Comment (RaMorey) [inline src/Conference/ConferenceDetails.jsx]: This is fine, but I'm wondering if it would be better to call it addFavouriteConference or something - I'm assuming in the future you might also implement a deleteFavouriteConference too? Although I guess you could call it unfavouriteConference

File: src/Conference/ConferenceDetails.jsx [hunk 2]
```diff
@@ -44,6 +44,14 @@ const ConferenceDetails = ({ conference, id, isLoading, error }) => {
...
+    } catch (error) {
+      console.log(error);
```
Comment (RaMorey) [inline src/Conference/ConferenceDetails.jsx]: Is there much use for this log? Might be nice to use the toast here for displaying the error

File: src/User/ProfilePage.jsx [hunk 3]
```diff
@@ -26,7 +60,21 @@ const ProfilePage = ({ user, isLoading, error }) => {
         <p className="email">Email: {email}</p>
         <p className="occupation">Occupation: {occupation}</p>
       </StyledCard>
-    </StyledForm>
+      <StyledCard>
+        <StyledCardHeading className="favourite-conferences">
+          Favourited conferences
+        </StyledCardHeading>
+        <StyledFavouritesList>
```
Comment (RaMorey) [inline src/User/ProfilePage.jsx]: I think it would be nice to pull this list out into its own component, helps seperate the logic

File: src/User/UserContainer.jsx [hunk 4]
```diff
@@ -1,18 +1,24 @@
...
-  const fetchData = async userId => {
+  const fetchData = async (userId, conferenceId) => {
     setIsLoading(true);
 
     try {
+      const favouriteConferences = await getFavouritedConferenceById(
+        conferenceId,
+        token
+      );
+      setFavouriteConferences(favouriteConferences);
```
Comment (RaMorey) [inline src/User/UserContainer.jsx]: It doesn't look like you're passing in the conferenceId in the useEffect?

File: src/api.js [hunk 5]
```diff
@@ -77,3 +77,24 @@ export const logoutUser = async token => {
...
+
+export const favouriteConference = async (conference, token) => {
+  try {
+    await axios.post("http://localhost:8080/userConferences", conference, {
+      headers: { Authorization: token }
+    });
+    toast.success("Conference has been favourited!");
+  } catch (error) {
+    toast.error("Conference has not been favourited!");
+    console.log(error);
```
Comment (RaMorey) [inline src/api.js]: You can remove the console log now you have the toast.

File: src/User/UserContainer.jsx [hunk 6]
```diff
@@ -1,18 +1,24 @@
...
   const [isLoading, setIsLoading] = useState(true);
   const [error, setError] = useState(false);
   const { id } = useParams();
 
-  const fetchData = async userId => {
+  const fetchData = async (userId, conferenceId) => {
     setIsLoading(true);
 
     try {
+      const favouriteConferences = await getFavouritedConferenceById(
```
Comment (RaMorey) [inline src/User/UserContainer.jsx]: I think either I'm confused or this should be userId? I assume you're trying to get a list of conferences a user has expressed interest in? But currently you're trying to get a list of conferences with conference ID

File: src/api.js [hunk 7]
```diff
@@ -77,3 +77,24 @@ export const logoutUser = async token => {
     throw error;
   }
 };
+
+export const getFavouritedConferenceById = async (id, token) => {
```
Comment (RaMorey) [inline src/api.js]: Could be nice to pluralise this to `getFavouritedConferencesById` or even `byUserId` to be more obvious what its doing

File: src/User/ProfilePage.jsx [hunk 8]
```diff
@@ -1,11 +1,45 @@
...
-  StyledCard
-} from "../StyledFormComponents";
-
-const ProfilePage = ({ user, isLoading, error }) => {
+import React, { useState, useEffect } from "react";
+import styled from "styled-components";
+import { Link } from "react-router-dom";
+import { StyledCardHeading } from "../StyledFormComponents";
+
+const StyledCard = styled.ul`
```
Comment (tonymaynard) [inline src/User/ProfilePage.jsx]: Don't you already have a styled card you could reuse?

File: src/Conference/ConferenceDetails.jsx (hunk 2 above)
Comment (sophiambolah) [reply src/Conference/ConferenceDetails.jsx]: Ok! I have a toast for this in the axios request, so should I just remove the console log and leave the catch part blank?

File: src/Conference/ConferenceDetails.jsx (hunk 1 above)
Comment (j-hutchinson) [reply src/Conference/ConferenceDetails.jsx]: I think (add/delete)favouriteConference or setFavouriteConference. Depends what the functionality is. Do you actively delete a favourite or just replace it?

File: src/Conference/ConferenceDetails.jsx (hunk 2 above)
Comment (RaMorey) [reply src/Conference/ConferenceDetails.jsx]: You can just remove the whole try catch and just write the `await favouriteConference(conference, token);` line I believe - let me know if it's working funny and I'll come over and help.

File: src/User/ProfilePage.jsx (hunk 8 above)
Comment (sophiambolah) [reply src/User/ProfilePage.jsx]: We have a reusable one for the add/edit conference and the conference details, but for the profile page I've changed the size dimensions so I don't think I can reuse it. I'll try changing the name of this card though so they're not the same!

Comment (tonymaynard) [reply src/User/ProfilePage.jsx]: Stlyed components are supposed to be pretty flexible, you could always pass props and set the size of the card that way instead of creating a new component with the exact same styling

File: src/Conference/FavouriteConferenceList.jsx [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  }
+`;
+
+const StyledConferenceLink = styled(Link)`
+  display: inline-block;
+  text-decoration: none;
+  color: #322d38;
+`;
+
+const FavouriteConferenceList = favouriteConferences => {
```
Comment (tonymaynard) [inline src/Conference/FavouriteConferenceList.jsx]: The first argument to a react component is the Props object, therefore you'll have to write it as ` ({ favouriteConferences })`

File: src/User/ProfilePage.jsx [hunk 10]
```diff
@@ -52,28 +32,20 @@ const ProfilePage = ({ user, favouriteConferences, isLoading, error }) => {
...
-          {favouriteConferences.map(conference => (
-            <StyledFavouritesListItem key={conference.id}>
-              <StyledConferenceLink to={`/${conference.id}`}>
-                {conference.name}
-              </StyledConferenceLink>
-            </StyledFavouritesListItem>
-          ))}
-        </StyledFavouritesList>
-      </StyledCard>
+        <FavouriteConferenceList />
```
Comment (tonymaynard) [inline src/User/ProfilePage.jsx]: You'll need to pass the list to component

**Comment by tonymaynard** [comment]
Is this PR still open/in progress?

File: src/App.js [hunk 11]
```diff
@@ -52,11 +51,11 @@ function App() {
             <Route path="/users/login">
               <LoginForm setSessionToken={setSessionToken} />
             </Route>
-            <Route path="/users">
+            <Route path="/users/:id">
               <UserContainer sessionToken={cookies.sessionToken} />
             </Route>
             <Route path="/:id">
-              <ConferenceContainer />
+              <ConferenceContainer sessionToken={cookies.sessionToken} />
```
Comment (tonymaynard) [inline src/App.js]: I don't think you need to pass the sessionToken prop

File: src/Conference/AddConferenceForm.jsx [hunk 12]
```diff
@@ -82,8 +84,9 @@ const Form = () => {
         required
         maxLength="20"
       />
-      <Input type="submit" value="submit" />
+      <Input type="submit" value="Submit" />
     </form>
   );
 };
-export default Form;
+const CookieAddConferenceForm = withCookies(AddConferenceForm);
```
Comment (tonymaynard) [inline src/Conference/AddConferenceForm.jsx]: This can just be writen as `export default withCookies(AddConferenceForm);`

Comment (tonymaynard) [reply src/Conference/AddConferenceForm.jsx]: This then allows you to import it in the other files with any name you want i.e. `AddConferenceForm` I wouldn't bother appending cookie at the front as the component doesn't care whether it needs cookies

File: src/Conference/AddConferenceForm.test.js [hunk 13]
```diff
@@ -1,51 +1,61 @@
+jest.mock("../api", () => ({
```
Comment (tonymaynard) [inline src/Conference/AddConferenceForm.test.js]: Out of interest why did you move this here? I thought they were hoisted by jest anyway so there wasn't a need to move them above imports

File: src/Conference/ConferenceContainer.jsx [hunk 14]
```diff
@@ -52,7 +52,7 @@ const ConferenceContainer = () => {
         />
       </Route>
       <Route path="/:id">
-        <ConferenceDetails
+        <CookieConferenceDetails
```
Comment (tonymaynard) [inline src/Conference/ConferenceContainer.jsx]: Don't think it should be prefixed with Cookie, Conference container doesn't care whether its involves cookies or not!

File: src/Conference/ConferenceDetails.jsx [hunk 15]
```diff
@@ -44,21 +53,38 @@ const ConferenceDetails = ({ conference, id, isLoading, error }) => {
...
         <p className="topic">{topic}</p>
         <p className="date">{getDatestring(date)}</p>
         <p className="time">{getTimestring(date)}</p>
         <p className="city">{city}</p>
         <StyledDescription className="description">
           {description}
         </StyledDescription>
-        <Button onClick={deleteThisConference}>Delete Conference</Button>
-      </StyledCard>
+        {token && (
```
Comment (tonymaynard) [inline src/Conference/ConferenceDetails.jsx]: Am I right in saying I could delete your conference? Is this expected?

File: src/Conference/AddConferenceForm.test.js (hunk 13 above)
Comment (sophiambolah) [reply src/Conference/AddConferenceForm.test.js]: I think Karam was told by someone that it was good practice to move the jest mocks to before the imports. We're happy to move them back though if they should be below?

File: src/Conference/UpdateConference.jsx [hunk 16]
```diff
@@ -1,12 +1,10 @@
 import React from "react";
-import Input from "../Input";
 import { useHistory } from "react-router-dom";
+import { withCookies, useCookies } from "react-cookie";
```
Comment (tonymaynard) [inline src/Conference/UpdateConference.jsx]: You're using a hook `useCookies` and a high order component `withCookies` in this component. You only need to use one

File: src/Conference/UpdateConference.jsx [hunk 17]
```diff
@@ -17,6 +15,8 @@ const UpdateForm = ({
   error
 }) => {
   const history = useHistory();
+  const [cookies] = useCookies([cookieName]);
+  const token = cookies.sessionToken.token;
```
Comment (tonymaynard) [inline src/Conference/UpdateConference.jsx]: Is there a possibility that someone who's not logged in lands on this route i.e. `conferences/1/edit`? therefore this sessionToken might be undefined and this may blow up. so might need to do a ternary check. Similarly I would look at doing a redirect if they land on this route in your conference container

File: src/Constants/Cookies.js [hunk 18]
```diff
@@ -0,0 +1,2 @@
+export const cookieName = "sessionToken";
```
Comment (tonymaynard) [inline src/Constants/Cookies.js]: This should probably be called sessionToken as its more descriptive but will cause a few files to change

File: src/Navbar.jsx [hunk 19]
```diff
@@ -1,6 +1,8 @@
 import React from "react";
-import styled from "styled-components";
+import { withCookies } from "react-cookie";
 import { Link, useHistory } from "react-router-dom";
+import { cookieOptions, cookieName } from "./Constants/Cookies";
+import styled from "styled-components";
```
Comment (tonymaynard) [inline src/Navbar.jsx]: It's convention to put absolute imports at the top

File: src/Navbar.jsx [hunk 20]
```diff
@@ -82,4 +91,5 @@ const Navbar = ({ sessionToken, deleteSessionToken }) => {
   );
 };
 
-export default Navbar;
+const CookieNavbar = withCookies(Navbar);
```
Comment (tonymaynard) [inline src/Navbar.jsx]: Can be written as `export default withCookies(Navbar);` don't need to assign it to a variable

File: src/Conference/ConferenceDetails.jsx (hunk 15 above)
Comment (sophiambolah) [reply src/Conference/ConferenceDetails.jsx]: Yes, at the moment we've built it so a logged-in user can add, delete and express interest in any conference. We don't currently have the functionality to delete only your own conferences, but I can add in another ticket for this!

File: src/StyledFormComponents.js [hunk 21]
```diff
@@ -17,12 +18,21 @@ const StyledCardHeading = styled.h3`
...
+  display: inline-block;
+  `
+      : `width: 900px;
+  height: 450px;
+  margin: 0 auto;`}
   border-style: solid;
   border-color: black;
   border-radius: 25px;
   background: #dbd8db;
-  width: 900px;
```
Comment (tonymaynard) [inline src/StyledFormComponents.js]: you could also write these as ` width: ${ ({ favouritesCard }) =>  favouritesCard ? '45%' : '900px' }` and same for the others!

File: src/StyledFormComponents.js [hunk 22]
```diff
@@ -32,4 +42,8 @@ const StyledCard = styled.div`
   position: relative;
 `;
 
-export { StyledCardHeading, StyledForm, StyledCard };
+const Card = ({ children, ...props }) => {
```
Comment (tonymaynard) [inline src/StyledFormComponents.js]: I don't think this component is needed, wherever you call this card I think you can write it as `<StyledCard {...props}>{ all your children needed } </StyledCard>` as it is already a component

File: src/User/LoginForm.jsx [hunk 23]
```diff
@@ -1,20 +1,23 @@
 import React, { useState } from "react";
 import styled from "styled-components";
 import { useHistory, Link } from "react-router-dom";
+import { withCookies, useCookies } from "react-cookie";
```
Comment (tonymaynard) [inline src/User/LoginForm.jsx]: Don't need both choose hook or HoC :)

**Comment by tonymaynard** [comment]
Try to keep PR's specific it seems theres lots of code changes here

File: src/Conference/AddConferenceForm.test.js (hunk 13 above)
Comment (tonymaynard) [reply src/Conference/AddConferenceForm.test.js]: No thats fine, I don't use jest so they probably know more than me :)

File: src/Conference/UpdateConference.jsx (hunk 17 above)
Comment (sophiambolah) [reply src/Conference/UpdateConference.jsx]: ok yes that's true! I'll keep a note of that for one of the clean up tickets

File: src/App.js (hunk 11 above)
Comment (tonymaynard) [reply src/App.js]: Did you mean to keep the sessionToken?

File: src/StyledFormComponents.js [hunk 24]
```diff
@@ -17,12 +17,14 @@ const StyledCardHeading = styled.h3`
 `;
 
 const StyledCard = styled.div`
+  width: ${({ favouritesCard }) => (favouritesCard ? `45%` : `900px`)};
+  height: ${({ favouritesCard }) => (favouritesCard ? `80vh` : `450px`)};
+  margin: ${({ favouritesCard }) => (favouritesCard ? `10px` : `0 auto`)};
+  display: ${({ favouritesCard }) => (favouritesCard ? `inline-block` : ``)};
```
Comment (tonymaynard) [inline src/StyledFormComponents.js]: This last one is probably better as  ${({ favouritesCard }) => favouritesCard && 'display: inline-block;'};

Comment (tonymaynard) [reply src/StyledFormComponents.js]: should remove the margin properity from line 33 too

File: src/User/ProfilePage.jsx [hunk 25]
```diff
@@ -32,15 +35,26 @@ const ProfilePage = ({ error, isLoading, user }) => {
...
-      <StyledLink className="editLink" to={`/users/edit/${id}`}>Edit</StyledLink>        
-      <StyledCardHeading className="name">
-        {`Hello, ${firstName} ${lastName}!`}
-      </StyledCardHeading>
-      <p className="username">{`Username: ${username}`}</p>
-      <p className="email">{`Email: ${email}`}</p>
-      <p className="occupation">{`Occupation: ${occupation}`}</p>
-    </StyledCard>
+    <StyledProfile>
+      <StyledCard favouritesCard>
```
Comment (tonymaynard) [inline src/User/ProfilePage.jsx]: should this be a favouritesCard?

File: src/App.js (hunk 11 above)
Comment (sophiambolah) [reply src/App.js]: oops sorry I forgot to remove this for the other components too!

File: src/User/UserContainer.jsx [hunk 26]
```diff
@@ -1,20 +1,29 @@
...
 import { Switch, Route, useParams } from "react-router-dom";
+import { useCookies } from "react-cookie";
+import { cookieName } from "../Constants/Cookies";
 import ProfilePage from "./ProfilePage";
-import { getUserById } from "../api.js";
 import UpdateProfile from "./UpdateProfile";
+import { getUserById, getFavouritedConferencesByUserId } from "../api.js";
 
-const UserContainer = ({ token }) => {
+const UserContainer = ({ sessionToken }) => {
```
Comment (tonymaynard) [inline src/User/UserContainer.jsx]: probably should remove this and get the cookie from the hook on line 15 to follow suit!

File: src/Navbar.jsx [hunk 27]
```diff
@@ -26,16 +28,19 @@ export const StyledLink = styled(Link)`
...
+  const sessionToken = allCookies.sessionToken;
   const logout = async () => {
     try {
       await logoutUser(sessionToken.token);
-      deleteSessionToken();
       history.push("/");
-    } catch (error) {
-      console.log(error);
-    }
+      deleteSessionToken();
```
Comment (tonymaynard) [inline src/Navbar.jsx]: Think this might complain that it was called before defined btw
